web/packages/teleport/src/JoinTokens/CreateJoinTokenDialog.tsx

```diff
@@ -0,0 +1,578 @@
+            <Text fontWeight={700} mb={2}>
+              AWS Account
+            </Text>
+
+            {index > 0 && ( // at least one rule is required, so lets not allow the user to remove it
+              <ButtonIcon onClick={() => removeRule(index)}>
+                <Trash size={16} color="text.muted" />
+              </ButtonIcon>
+            )}
+          </Flex>
+          <FieldInput
+            label="Account ID"
+            rule={requiredField('Account ID is required')}
+            placeholder="aws account ID"
+            value={rule.aws_account}
+            onChange={e =>
+              setTokenRulesField(index, 'aws_account', e.target.value)
+            }
+          />
+          <FieldInput
+            label="ARN"
+            toolTipContent={`The joining nodes must match this ARN. Supports wildcards "*" and "?"`}
+            placeholder="AWS ARN"
+            value={rule.aws_arn}
+            onChange={e => setTokenRulesField(index, 'aws_arn', e.target.value)}
+          />
+        </RuleBox>
+      ))}
+      <ButtonText onClick={addNewRule}>
+        <Plus size={16} mr={2} />
+        Add another AWS Account
+      </ButtonText>
+    </>
+  );
+};
+
+type OptionGCP = Option<string, string>;
+
+const JoinTokenGCPForm = ({
+  tokenState,
+  onUpdateState,
+}: {
+  tokenState: NewJoinTokenState;
+  onUpdateState: (newToken: NewJoinTokenState) => void;
+}) => {
+  const rules = tokenState.gcp;
+  function removeRule(index: number) {
+    const newRules = [...rules].filter((rule, i) => {
+      if (index === i) {
+        return false;
+      }
+      return rule;
+    });
+    const newState = {
+      ...tokenState,
+      gcp: newRules,
+    };
+    onUpdateState(newState);
+  }
+
+  function addNewRule() {
+    const newState = {
+      ...tokenState,
+      gcp: [
+        ...tokenState.gcp,
+        { project_ids: [], locations: [], service_accounts: [] },
+      ],
+    };
+    onUpdateState(newState);
+  }
+
+  function updateRuleField(
+    index: number,
+    fieldName: string,
+    opts: OptionGCP[]
+  ) {
+    const newState = {
+      ...tokenState,
+      gcp: tokenState.gcp.map((rule, i) => {
+        if (i === index) {
+          return { ...rule, [fieldName]: opts };
+        }
+        return rule;
+      }),
+    };
+    onUpdateState(newState);
+  }
+
+  return (
+    <>
+      {rules.map((rule, index) => (
+        <RuleBox
+          key={index} // order doesn't change without updating the referrenced array
+        >
+          <Flex alignItems="center" justifyContent="space-between">
+            <Text fontWeight={700} mb={2}>
+              GCP Rule
+            </Text>
+
+            {index > 0 && ( // at least one rule is required, so lets not allow the user to remove it
+              <ButtonIcon onClick={() => removeRule(index)}>
+                <Trash size={16} color="text.muted" />
+              </ButtonIcon>
+            )}
```

Comment:
It would be a better UX if you displayed the delete button if `rules.length > 1` instead of `index > 0`

Reply:
Oh i guess i see your point, because we may want to delete the first rule and use the next ones. I didn't even think of that. genius